src/bp/bpfs.ts

```diff
@@ -0,0 +1,229 @@
+    this.targetDir = args.targetDir && path.resolve(args.targetDir)
+    this.sourceDir = args.sourceDir && path.resolve(args.sourceDir)
+
+    if (!this.serverUrl || !this.authToken || !this.authToken.length) {
+      this._endWithError(`Missing parameter "url" or "authToken"`)
+    }
+
+    if (action === 'pull' && !this.targetDir) {
+      this._endWithError(`Target directory is not valid: "${this.targetDir}"`)
+    }
+
+    if (action === 'push' && !this.sourceDir) {
+      this._endWithError(`Source directory must be set: "${this.sourceDir}"`)
+    }
+  }
+
+  async pullChanges() {
+    const cleanBefore = process.argv.includes('--clean')
+    const axiosClient = this._getPullAxiosClient()
+    const dryRun = process.argv.includes('--dry')
+
+    try {
+      // We clear only those two folders, so assets are preserved
+      if (cleanBefore) {
+        console.log(chalk.blue(`Cleaning data folder before pulling data...`))
+        await this._clearDir(path.join(this.targetDir, 'global'))
+        await this._clearDir(path.join(this.targetDir, 'bots'))
+      } else if (fse.existsSync(this.targetDir)) {
+        const fileCount = await this._filesCount(this.targetDir)
+        console.log(chalk.blue(`Remote files will be pulled in an existing folder containing ${fileCount} files`))
+      }
+
+      console.log(chalk.blue(`Pulling all remote changes from ${this.serverUrl} to ${this.targetDir} ...`))
+
+      const { data: archive } = await axiosClient.get('export')
+      if (dryRun) {
+        console.log(chalk.yellow(`Dry run completed. Successfully downloaded archive.`))
+        return process.exit()
+      }
+
+      console.log(
+        chalk.blue(`Extracting archive to local file system... (archive size: ${bytesToString(archive.length)}`)
+      )
+
+      await extractArchive(archive, this.targetDir)
+
+      const fileCount = await this._filesCount(this.targetDir)
+      console.log(chalk.green(`Successfully extracted ${fileCount} files from the remote server!`))
+    } catch (err) {
+      const error = err.response ? `${err.response.statusText} (${err.response.status})` : err.message
+      this._endWithError(`Could not pull changes: ${error}`)
+    }
+  }
+
+  async pushChanges() {
+    const useForce = process.argv.includes('--force')
+    const dryRun = process.argv.includes('--dry')
+
+    if (!fse.existsSync(this.sourceDir)) {
+      this._endWithError(`Specified folder "${this.sourceDir}" doesn't exist.`)
+    }
+
+    try {
+      console.log(chalk.blue(`Preparing an archive of your local files from ${this.sourceDir}...`))
+
+      const archive = await createArchiveFromFolder(this.sourceDir, pushedArchiveIgnoredFiles)
+      const axiosClient = this._getPushAxiosClient(archive.length)
+
+      console.log(
+        chalk.blue(`Sending archive to server for comparison... (archive size: ${bytesToString(archive.length)})`)
+      )
+      const { data } = await axiosClient.post('changes', archive)
```

Comment:
I have a small suggestion for you. Instead of wrapping your push client in a "factory method" that you use at a single place then use it to post and receive data without typings, you could wrap your http call in the method and mark the types, maybe use `processChanges` within this function as well. Here's what I mean

```ts
// ... 
  const archive = await createArchiveFromFolder(this.sourceDir, pushedArchiveIgnoredFiles)

  console.log(chalk.blue(`Sending archive to server for comparison... (archive size: ${bytesToString(archive.length)})`)

  const { changeList, blockingChanges, localFiles } = this._pushArchive(archive)

/*
 further in the file
*/

private _postArchive(archive: Buffer) : {changeList: string[], blockingChanges: change[], localfiles: string[]} => { /* . .. */}
```